cluster-autoscaler/provisioningrequest/checkcapacity/provisioningclass.go

```diff
@@ -95,130 +95,123 @@ func (o *checkCapacityProvClass) Provision(
 	}
 
-	return combinedStatus.Export()
+	batch, err := o.provreqInjector.GetCheckCapacityBatch(o.checkCapacityProvisioningRequestMaxBatchSize)
+	klog.Infof("Processing provisioning requests as batch of size %d", len(batch))
+	return batch, err
 }
 
-// checkcapacity checks if there is capacity in the cluster for pods from a ProvisioningRequest. It persists the scheduling in cluster snapshot whenever all pods can be scheduled, and doesn't change the snapshot otherwise.
-func (o *checkCapacityProvClass) checkcapacity(unschedulablePods []*apiv1.Pod, provReq *provreqwrapper.ProvisioningRequest) (bool, error) {
-	var capacityAvailable bool
-	err, cleanupErr := clustersnapshot.WithForkedSnapshot(o.context.ClusterSnapshot, func() (bool, error) {
-		st, _, err := o.schedulingSimulator.TrySchedulePods(o.context.ClusterSnapshot, unschedulablePods, scheduling.ScheduleAnywhere, true)
-		if len(st) < len(unschedulablePods) || err != nil {
-			if noRetry, ok := provReq.Spec.Parameters[NoRetryParameterKey]; ok && noRetry == "true" {
-				// Failed=true condition triggers retry in Kueue. Otherwise ProvisioningRequest with Provisioned=Failed
-				// condition block capacity in Kueue even if it's in the middle of backoff waiting time.
-				conditions.AddOrUpdateCondition(provReq, v1.Failed, metav1.ConditionTrue, conditions.CapacityIsNotFoundReason, "CA could not find requested capacity", metav1.Now())
-			} else {
-				if noRetry, ok := provReq.Spec.Parameters[NoRetryParameterKey]; ok && noRetry != "false" {
-					klog.Errorf("Ignoring Parameter %v with invalid value: %v in ProvisioningRequest: %v. Supported values are: \"true\", \"false\"", NoRetryParameterKey, noRetry, provReq.Name)
-				}
-				conditions.AddOrUpdateCondition(provReq, v1.Provisioned, metav1.ConditionFalse, conditions.CapacityIsNotFoundReason, "Capacity is not found, CA will try to find it later.", metav1.Now())
-			}
-			capacityAvailable = false
-		} else {
-			conditions.AddOrUpdateCondition(provReq, v1.Provisioned, metav1.ConditionTrue, conditions.CapacityIsFoundReason, conditions.CapacityIsFoundMsg, metav1.Now())
-			capacityAvailable = true
-		}
+func (o *checkCapacityProvClass) isBatchEnabled() bool {
+	return o.provreqInjector != nil && o.checkCapacityProvisioningRequestMaxBatchSize > 1
+}
 
-		_, updErr := o.client.UpdateProvisioningRequest(provReq.ProvisioningRequest)
-		if updErr != nil {
-			return false, fmt.Errorf("failed to update Provisioned condition to ProvReq %s/%s, err: %v", provReq.Namespace, provReq.Name, updErr)
+func (o *checkCapacityProvClass) checkCapacityBatch(reqs []provreq.ProvisioningRequestWithPods, combinedStatus *combinedStatusSet, startTime time.Time) []*provreqwrapper.ProvisioningRequest {
+	updates := make([]*provreqwrapper.ProvisioningRequest, 0, len(reqs))
+	for _, req := range reqs {
+		// timebox checkCapacity when batch processing.
+		if o.isBatchEnabled() && time.Since(startTime) > o.checkCapacityProvisioningRequestBatchTimebox {
```

Comment:
Shouldn't the first iteration be skipped in this check?

Reply:
That sounds reasonable. I moved the check to the end of the loop.